src/Components/Components/src/RenderTree/Renderer.cs

```diff
@@ -634,11 +635,31 @@ private void RenderInExistingBatch(RenderQueueEntry renderQueueEntry)
                 var disposeComponentId = _batchBuilder.ComponentDisposalQueue.Dequeue();
                 var disposeComponentState = GetRequiredComponentState(disposeComponentId);
                 Log.DisposingComponent(_logger, disposeComponentState);
-                if (!disposeComponentState.TryDisposeInBatch(_batchBuilder, out var exception))
+                if (!disposeComponentState.RequiresAsyncDisposal())
+                {
+                    if (!disposeComponentState.TryDisposeInBatch(_batchBuilder, out var exception))
+                    {
+                        exceptions ??= new List<Exception>();
+                        exceptions.Add(exception);
+                    }
+                }
+                else
                 {
-                    exceptions ??= new List<Exception>();
-                    exceptions.Add(exception);
+                    var result = disposeComponentState.DisposeInBatchAsync(_batchBuilder);
+                    if (result.IsCompleted)
+                    {
+                        if (result.IsFaulted && !result.IsCanceled)
+                        {
+                            exceptions ??= new List<Exception>();
+                            ProcessAggregateException(exceptions, result.Exception);
+                        }
+                    }
+                    else
+                    {
+                        _ = GetErrorHandledTask(result);
```

Comment:
`GetErrorHandledTask` ignores cancellation, so if we do choose to make this an error out of caution, then we'd need a different approach. Either passing an extra flag to `GetErrorHandledTask`, or if we're super perf-focused, having a second method like `GetErrorHandledTaskDisallowingCancellation` that omits the cancellation check.

Reply:
Also, shouldn't you call `AddToPendingTasks` with the resulting task here? Otherwise what's the intended behavior for prerendering - shouldn't it wait for the async disposal tasks and fail if any of the fail?

Reply:
This is gone now, I added the task to the list of pending tasks, and wrote an E2E test to ensure we wait for async disposal to be completed.